src/box/lua/schema.lua

```diff
@@ -213,6 +214,28 @@ box.priv = {
     ["ALL"] = builtin.PRIV_ALL
 }
 
+local error_t = ffi.typeof('struct error')
+
+--
+-- Call fn with given variable number of arguments. In case of error
+-- if error is box.error then update trace frame to the given level.
+--
+local function call_set_trace(level, fn, ...)
+    local st, res = pcall(fn, ...)
+    if not st then
+        if ffi.istype(error_t, res) then
+            box.error(res, advance_trace(level))
+        else
+            -- Keep original trace. It is useful in case of programmatic
+            -- errors.
+            error(res)
+        end
+    end
+    return res
+end
```

Comment:
- This function should be defined in the error module. It's probably better be named `box.error.internal.call_at`.
- I think you could use the tail call optimization here to avoid advancing the level.
- The function may return more than one value on success. You should probably handle this case.

Reply:
> * This function should be defined in the error module. It's probably better be named `box.error.internal.call_at`.

Come to think of it, let's move it to the utils module.

Reply:
Done. But decided not to use tail call optimization after discussion f2f. It is uncommon for raising an error.